### PR TITLE
Parallelization of semantic tokens

### PR DESCRIPTION
**Pull request by moosq**

### CODE REVIEW

File: internal/lsp/semantic_tokens.go [hunk 1]
```diff
@@ -172,6 +177,64 @@ var builtins = map[string]SemanticTokenType{
...
+	Begin loc.Pos
+	End   loc.Pos
+}
+
+type SemTokSeqItem struct {
+	Data []uint32
+	Idx  int
+}
+
+func LastTokenLine(d *[]uint32) uint32 {
```
Comment (5nord) [inline internal/lsp/semantic_tokens.go:190]: I think you should use the slice directly and not a pointer to it.

File: internal/lsp/semantic_tokens.go [hunk 2]
```diff
@@ -172,6 +177,64 @@ var builtins = map[string]SemanticTokenType{
...
+	var sum uint32 = 0
+	for i := 0; i < len(*d); i += 5 {
+		sum += (*d)[i]
+	}
+	return sum
+}
+
+func SemanticTokenReassambly(seqItems []SemTokSeqItem) []uint32 {
```
Comment (5nord) [inline internal/lsp/semantic_tokens.go:198]: I would prefer a more general implementation. For example:
```
func AppendDelta(s []uint32) []uint32
```

File: internal/lsp/semantic_tokens.go [hunk 3]
```diff
@@ -172,6 +177,64 @@ var builtins = map[string]SemanticTokenType{
...
+			nextEndL := begin.Line + r/int(nrOfRanges)
+			if i == int(nrOfRanges-1) {
+				res = append(res, Range{tree.Pos(begin.Line, begin.Column), e})
+			} else {
+				res = append(res, Range{tree.Pos(begin.Line, begin.Column), tree.Pos(nextEndL, 1) - 1})
+			}
+			begin.Column = 1
+			begin.Line += r / int(nrOfRanges)
+		}
+	} else {
```
Comment (5nord) [inline internal/lsp/semantic_tokens.go:231]: It's considered good practice to do without else clause by early return. For example:
```go
r := end.Line - begin.Line
if r <= splitAfterlines {
       return Range{b, e}
}
for i := 0; i < nrOfRangees; i++ {
```

But in this case you could move the condition to the caller even?

File: internal/lsp/semantic_tokens_test.go [hunk 4]
```diff
@@ -518,19 +589,54 @@ func testTokens(t *testing.T, rng *protocol.Range, text string) []Token {
...
 	)
-	list := lsp.SemanticTokens(tree, db, begin, end)
 	for i := 0; i < len(list); i += 5 {
 		line += int(list[i])
 		if list[i] != 0 {
 			col = 1
 		}
 		col += int(list[i+1])
-		pos := int(tree.Pos(line, col)) - 1 // -1 to account for the Pos offset.
+		pos := int(tree.Pos(line, col)) - 1
```
Comment (5nord) [inline internal/lsp/semantic_tokens_test.go:639]: Why remove the comment?

Comment (moosq) [reply internal/lsp/semantic_tokens_test.go]: I restored it

**Comment by moosq** [comment]
`tree.ParentOf` holds a mutex which is needed to protect filling of the Node map. This makes parallelization very ineffective. A better solution might render parallelization of semantic tokens as superfluous

**Pull request closed without merging**